substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/JavaMonitor.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2022, 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2022, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.monitor;
+
+import java.util.LinkedList;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.graalvm.nativeimage.CurrentIsolate;
+import com.oracle.svm.core.jfr.SubstrateJVM;
+import com.oracle.svm.core.jfr.events.JavaMonitorEnterEvent;
+import com.oracle.svm.core.jfr.JfrTicks;
+
+public class JavaMonitor extends ReentrantLock {
+    private long ownerTid;
+
+    //Records the number of waiters that have not yet been notified
+    private int waitingThreads = 0;
+    /*
+     * This queue is used to record the TIDs of notifiers so that waiters have access to them when they eventually
+     * acquire the monitor lock and resume.
+     * Adding and removing TIDs from this queue in a FIFO fashion is correct because Condition.signal(), notifies waiters in a FIFO fashion.
```

Comment:
I don't think that this assumption is correct because:
a. it is possible to specify a timeout (see `wait(long)`)
b. it is possible to interrupt a thread

Reply:
a. If a thread wakes up from a timeout then it will not pull anything off of the TID queue. It will instead just emit an event with the `timedout` flag set. [see here: ](https://github.com/oracle/graal/pull/4676/files#diff-323fb7546644875e073e4f7271cbc476ef672e237278eff7e21f7d8d47d928a8R383-R387)

b. If a waiting thread is interrupted it will also not affect the TID queue. 

I believe this means that a waiter will modify the notifier TID queue if and only if it has been notified.

Reply:
Thank you Christian. I have changed things so that we always clean up after a waiter resumes (through timeout, interrupt, or signal).  I have also changed things so that waiters are checked to still be waiting before we mark them as being notified (so that notifications don't get associated with waiters that have timed out or interrupted).  Some drawbacks of this approach  are that a linear scan of the maintained queue is required when waiters wake up. This is because the arbitrary order in which threads reacquire the lock. 

In Hotspot, a notifierTid field is added to the Node struct that sits on the wait and sync queues. This way, when a waiter wakes up, it can simply inspect that field. However, in substrateVM, the wait queue and sync queue is abstracted away in `Condition.signal() and Condition.await()` the solution in this PR tries to mirror  this with its own queue. I have created another PR [here](https://github.com/oracle/graal/pull/4698), that uses substitutions to copy the behaviour in Hotspot more closely. This alternative approach is much simpler, but requires maintenance of the substituted methods when the original methods change.